rfcs/2020-07-21-2744-vicscript.md

```diff
@@ -0,0 +1,191 @@
+```
+
+As well as a timestamp type:
+
+```coffee
+created_at = now().unix()
+```
+
+Boolean operators and arithmetic galore:
+
+```coffee
+is_big = number > 100
+multiplied = number * 7
+```
+
+Perform assignments conditionally with an `if` statement:
+
+```coffee
+sorted_foo = if foo.type() == "array" { foo.sort() }
+```
+
+And use a `match` statement for pattern matching:
+
+```coffee
+new_doc = match doc {
+  type == "article" => article
+  type == "comment" => comment
+  _ => this
+}
+```
+
+Use a wealth of methods on values in order to perform common mutations on them:
+
+```coffee
+sorted = foo.sort()
+uppercase = bar.uppercase()
+
+foo = bar.map_each(if this.description.contains("delete me") {
+    deleted()
+} else {
+    this
+})
+```
+
+And, finally, create your own re-usable maps with the `map` keyword:
+
+```coffee
+map things {
+  first  = thing_one
+  second = thing_two
+}
+
+foo = value_one.apply("things")
+bar = value_two.apply("things")
+```
+
+## Examples
+
+I've written an implementation of [an embedded CSV parser](2020-07-21-2744-vicscript/example1.coffee) using the above proposal in order to demonstrate what it looks like. This example is currently runnable with Benthos, and I've used the `.coffee` suffix because the syntax highlighting for CoffeeScript basically gives us everything we need already.
+
+## Prior Art
+
+### JQ
+
+JQ is a great tool and rightfully gets a lot of love. It also basically solves the same problem as Vicscript. Unfortunately, it's common knowledge that JQ doesn't score high on readability, and therefore it doesn't scale well at all in relation to mapping size and complexity (imagine trying to solve [https://github.com/timberio/vector/issues/1965](https://github.com/timberio/vector/issues/1965) with it).
+
+The modules syntax introduced in 1.5 helps a lot by allowing you to break your mapping down, but the syntax is still difficult to learn and awkward to read.
+
+### Tremor Script
+
+[Tremor Script](https://docs.tremor.rs/tremor-script/) is part of the [Tremor](https://docs.tremor.rs/) project and therefore is closely aligned with Vector. Of all existing alternatives this seems the most likely candidate for quick adoption as it's written in Rust and basically solves the same problem we have.
+
+Tremor script is (obviously) designed to work with records as they're modelled within Tremor, we therefore might struggle to get it working with our own event types and the translations to/from may have a performance penalty.
+
+It also seems like there are some key limitations with the language that we would need to contribute solutions for (or fork). The main one being simple coalescing expressions, I can't find anything in their docs to allow something like `foo = bar.(baz | buz).bev` without a nasty bunch of nested `match` expressions.
+
+Similar to JQ the syntax is also far enough removed from basic c-style declarations that there's a learning curve and more difficulty in maintaining syntax highlighters.
```

Comment:
Agree with all of the above regarding Tremor script. I think it's an impressive feature of Tremor, but would like to avoid it for the reasons you listed.